drivers/node/vsphere/vsphere.go

```diff
@@ -862,3 +862,336 @@ func (v *vsphere) RemoveNonRootDisks(n node.Node) error {
+	}
+
+	sourceDatastore := object.NewDatastore(vm.Client(), largestDisks[0].Datastore)
+
+	targetDatastores, err = filterTargetDatastores(ctx, sourceDatastore, compatibleDatastores)
+	if err != nil {
+		return fmt.Errorf("error filtering target datastores: %v", err)
+	}
+
+	if !moveAllDisks {
+		log.Infof("Trying to Move largest disk on VM %v to Datastore : %v", node.Name, targetDatastores[0].Reference().Value)
+		err = initiateStorageVmotion(ctx, vm, largestDisks[:1], targetDatastores)
+		if err != nil {
+			return fmt.Errorf("error during storage vMotion: %v", err)
+		}
+	} else {
+		log.Infof("Trying to Move all disks from %v to Datastore : %v", node.Name, targetDatastores[0].Reference().Value)
+		diskLocators := make([]types.VirtualMachineRelocateSpecDiskLocator, 0)
+		for _, device := range vmProps.Config.Hardware.Device {
+			if disk, ok := device.(*types.VirtualDisk); ok {
+				diskLocators = append(diskLocators, types.VirtualMachineRelocateSpecDiskLocator{
+					DiskId:    disk.Key,
+					Datastore: targetDatastores[0].Reference(),
+				})
+			}
+		}
+		if len(diskLocators) == 0 {
+			return fmt.Errorf("no disks found on the VM")
+		}
+		err = initiateStorageVmotion(ctx, vm, diskLocators, targetDatastores)
+		if err != nil {
+			return fmt.Errorf("error during storage vMotion: %v", err)
+		}
+		postData, err := GetCloudDriveConfigmapData(stc)
+		if err != nil {
+			return fmt.Errorf("error fetching post-vMotion cloud drive config: %v", err)
+		}
+		if !v.ValidateDatastoreUpdate(preData, postData, node.VolDriverNodeID, targetDatastores[0].Reference().Value) {
+			return fmt.Errorf("validation failed: datastore updates are not as expected")
+		}
+	}
+	return nil
+}
+
+// Function to get the datastore's cluster
+func getDatastoreCluster(ctx context.Context, ds *object.Datastore) (*object.StoragePod, error) {
+	var dsProps mo.Datastore
+	err := ds.Properties(ctx, ds.Reference(), []string{"parent"}, &dsProps)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get properties for datastore: %v", err)
+	}
+
+	if dsProps.Parent != nil {
+		spRef := dsProps.Parent.Reference()
+		if spRef.Type == "StoragePod" {
+			return object.NewStoragePod(ds.Client(), spRef), nil
+		}
+	}
+
+	return nil, nil
+}
+
+// filterTargetDatastores filters from list of Datastores available with the prefix defined in Storageclass
+// If source DS is in a cluster with another DS, then this method skips the other DS and chooses a DS outside this cluster
+func filterTargetDatastores(ctx context.Context, sourceDatastore *object.Datastore, allDatastores []*object.Datastore) ([]*object.Datastore, error) {
+	sourceCluster, err := getDatastoreCluster(ctx, sourceDatastore)
+	if err != nil {
+		return nil, err
+	}
+	var filteredDatastores []*object.Datastore
+
+	for _, ds := range allDatastores {
+		if ds.Reference().Value == sourceDatastore.Reference().Value {
+			continue
+		}
+		targetCluster, err := getDatastoreCluster(ctx, ds)
```

Comment:
It is better to consider each datastore in datastore cluster as a separate datastore. So that we can svmotion between datastores of a single datastorecluster.

Reply:
Yes that's happening even with this code